src/Spectre.libGaussianMixtureModelling/ExpectationRunnerOpt.h

```diff
@@ -0,0 +1,113 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#pragma once
+#include <vector>
+#include "DataTypes.h"
+#include "GaussianDistribution.h"
+#include "GaussianMixtureModel.h"
+#include "Matrix.h"
+
+namespace spectre::unsupervised::gmm
+{
+    /// <summary>
+    /// Class serves the purpose of expectation step of Expectation
+    /// Maximization algorithm. It is partially optimized, but still uses
+    /// just a single thread to compute.
+    /// </summary>
+    class ExpectationRunnerOpt
+    {
+    public:
+        /// <summary>
+        /// Constructor initializing the class with data required during
+        /// expectation step.
+        /// </summary>
+        /// <param name="spectrum">Spectrum to operate on.</param>
+        /// <param name="affilationMatrix">Matrix symbolising the probability
+        /// of affilation of each sample to a certain gaussian component.
+        /// </param>
+        /// <param name="components">Gaussian components.</param>
+        ExpectationRunnerOpt(SpectrumView spectrum,
+            Matrix<DataType> &affilationMatrix,
+            std::vector<GaussianComponent> &components)
+            : m_Spectrum(spectrum), m_Components(components)
+            , m_AffilationMatrix(affilationMatrix)
+            , m_Sums(spectrum.mzs.size())
+        {
+        }
+
+        /// <summary>
+        /// Fills affilation (gamma) matrix with probabilities of affilation
+        /// of each sample to a certain gaussian component.
+        /// </summary>
+        void Expectation()
+        {
+            DataView mzs = m_Spectrum.mzs;
+            DataView intensities = m_Spectrum.intensities;
+            std::fill(m_Sums.begin(), m_Sums.end(), 0.0);
+            const unsigned numberOfComponents = (unsigned)m_Components.size();
+            for (unsigned k = 0; k < numberOfComponents - 1; k++)
+            {
+                for (unsigned i = 0; i < (unsigned)mzs.size(); i++)
+                {
+                    DataType probability = m_Components[k].weight *
+                        Gaussian(mzs[i], m_Components[k].mean,
+                            m_Components[k].deviation);
+                    m_AffilationMatrix[k][i] = probability;
+                    m_Sums[i] += probability;
+                }
+            }
+            const unsigned lastK = numberOfComponents - 1;
+            for (unsigned i = 0; i < (unsigned)mzs.size(); i++)
+            {
+                DataType probability = m_Components[lastK].weight *
+                    Gaussian(mzs[i], m_Components[lastK].mean,
+                        m_Components[lastK].deviation);
+                m_AffilationMatrix[lastK][i] = probability;
+                m_Sums[i] = 1.0 / (m_Sums[i] + probability);
```

Comment:
Why does this differ from the version above? Could you add any comment or whatever?

Reply:
The formula that is computed in this function is represented by:
![image](https://user-images.githubusercontent.com/4967505/40992964-87ec0c70-68f8-11e8-99e4-630531869e15.png)
Since the divison is a very costly operation, I decided to minimize it in such a way:
1. Compute the sum of probabilities for each component from j=0 up to K=1 and save it in sums[i]
2. For the last component, add the final probability and take an inverse of the sums.
3. That way, in the loop below, we no longer have to perform division for each `K*N` elements, but we can apply a much more efficient multiplication. That way, we save ourselves (K-1)*N division operations.